### ISSUE TITLE
Documentation: use ORM declarative attributes in .values instead of kwargs in "ORM-enabled UPDATE statements"?

### ISSUE THREAD

**Issue by MaicoTimmerman** [sqlalchemy/sqlalchemy#6332]
**Describe your question**
I noticed that in  documentation for ORM-enabled UPDATE statement mention the keyword argument based `.values()`. Shouldn't these places be used to show the ORM style `.values`? Now it uses the same methodology as the core expression language.

For example:
```py
>>> session.execute(
...     update(User).
...     where(User.name == "sandy").
...     # values(fullname="Sandy Squirrel Extraodinaire") # not this
...     values({User.fullname: "Sandy Squirrel Extraodinaire"})
... )
```

I think it applies in the [1.4 tutorial](https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html?highlight=update#orm-enabled-update-statements), [the session basics](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#update-and-delete-with-arbitrary-where-clause) and perhaps in more place?

**Comment by zzzeek** [comment]
im not following.  are you saying replace

    update(User).where(User.name == "squidward").values(name="spongebob")

with

    update(User).where(User.name == "squidward").values({User.name: "spongebob"})

?

**Comment by MaicoTimmerman** [comment]
Yes, as the second pattern matches better with the ORM declarative syntax.

**Comment by zzzeek** [comment]
the first version is much more convenient for most use cases.    I think the ORM query.update() doesn't offer this, which is a shortcoming.

**Comment by zzzeek** [comment]
if the issue is, you're going for good mypy support, that should just be discussed generially w.r.t the update() construct.  someone who reads the tutorial should know that both forms are available.    I can attest that keyword arg forms for methods like these remain very popular.

**Comment by MaicoTimmerman** [comment]
Mypy support is indeed a big +1, I thinkan update to the stubs is enough for that, I'll try and work on a PR for that.

I found the pattern while looking through the tests and couldn't find them in the docs to refer to my peers. Should it be expressed in the docs?

**Comment by zzzeek** [comment]
it should likely be at https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Update.values  and we may additionally refer to it at https://docs.sqlalchemy.org/en/14/tutorial/data.html#tutorial-core-update-delete as well.

mentioning it at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#update-and-delete-with-arbitrary-where-clause is something to consider but that might make this section more crowded which it already sort of is, maybe a new sub-section, because up until now, nobody would have preferred `{User.name: value}` except for IDE/mypy support, so it's likely going to be a 50/50 thing going forward.

theoretically, the mypy plugin should even know how to set up the correct **kwargs for the values() method.   That's actually a thing that could work (but wouldn't help the IDE case).

The `{expr: value}` format is not common because the only use case where it's necessary is the very unusual case of updating the specific index of ARRAY or JSON, which is a PostgreSQL only thing.  So the syntax seems to only be mentioned at https://docs.sqlalchemy.org/en/14/core/type_basics.html?highlight=array#sqlalchemy.types.ARRAY .

**Comment by MaicoTimmerman** [comment]
I'll have to experiment with some more cases, but I think it should work with something like:

```py
class ValuesBase(UpdateBase):
    ...
    @overload
    def values(self: _VB, **kwargs: Any) -> _VB: ...
    @overload
    def values(self: _VB, argument: Mapping[_T, _T]) -> _VB: ...

class Insert(ValuesBase):
    ...
    @overload
    def values(self: _VB, **kwargs: Any) -> _VB: ...
    @overload
    def values(self: _VB, argument: Mapping[_T, _T]) -> _VB: ...
    @overload
    def values(self: _VB, argument: List[Mapping[_T, _T]]) -> _VB: ...
```
where the plugin could narrow `**kwargs` down.

> The `{expr: value}` format is not common because the only use case where it's necessary is the very unusual case of updating the specific index of ARRAY or JSON, which is a PostgreSQL only thing. So the syntax seems to only be mentioned at https://docs.sqlalchemy.org/en/14/core/type_basics.html?highlight=array#sqlalchemy.types.ARRAY .

I did not consider the full expression case yet, though. I'll work on that later and also add some test cases to `sqlalchemy/test/ext/mypy/files/`.

**Comment by zzzeek** [comment]
is this a plugin issue or a general typing issue?   im planning on moving forward without the plugin, which will become legacy.  for 2.0 it's better to slightly modify APIs to be more pep-484 friendly so that there aren't really any "bugs" on our end.

**Comment by MaicoTimmerman** [comment]
I would say a general typing issue. When `.values()` accept a mapping with declarative keys, it can become fully typed. The kwargs-based arguments will not be type-checkable to my knowledge.

**Comment by zzzeek** [comment]
can we clarify what we want to do in this issue?    2.0 is fully typed now including the dictionary sent to `values()`.

**Comment by chen-tamarin** [comment]
```python
>>> session.execute(
...     update(User).
...     where(User.name == "sandy").
...     # values(fullname="Sandy Squirrel Extraodinaire") # not this
...     values({User.fullname: "Sandy Squirrel Extraodinaire"})
... )
```

This works in 2.0 now, thank you! The docs don't show this capability though: https://docs.sqlalchemy.org/en/20/core/dml.html#sqlalchemy.sql.expression.Insert.values.params.*args
